### PR TITLE
Ftr/dev 6005 add cfda details endpoint

### PR DESCRIPTION
**Pull request by BrianZito**
**Description:**
Adds reference endpoint to supply further details about CFDAs from the grants API.

**Technical details:**

**Requirements for PR merge:**

1. [ ] Unit & integration tests updated
2. [ ] API documentation updated
3. [ ] Necessary PR reviewers:
    - [ ] Backend
    - [x] Frontend <OPTIONAL>
    - [ ] Operations <OPTIONAL>
    - [ ] Domain Expert <OPTIONAL>
4. [ ] Matview impact assessment completed
5. [ ] Frontend impact assessment completed
6. [ ] Data validation completed
7. [ ] Appropriate Operations ticket(s) created
8. [x] Jira Ticket [DEV-6005](https://federal-spending-transparency.atlassian.net/browse/DEV-6005):
    - [x] Link to this Pull-Request
    - [ ] Performance evaluation of affected (API | Script | Download)
    - [ ] Before / After data comparison

**Area for explaining above N/A when needed:**
```
```

### CODE REVIEW

**Comment by BrianZito** [comment]
"Do not merge" added since whitelisting has not been testing in Sandbox yet.

File: usaspending_api/api_contracts/contracts/v2/references/cfda/totals/cfda.md
Comment (tony-sappe) [inline usaspending_api/api_contracts/contracts/v2/references/cfda/totals/cfda.md:44]: Is null a valid CFDA?

Comment (BrianZito) [reply usaspending_api/api_contracts/contracts/v2/references/cfda/totals/cfda.md]: it is not; corrections have been made

File: usaspending_api/references/v2/views/cfda.py
Comment (tony-sappe) [inline usaspending_api/references/v2/views/cfda.py:17]: No API cache?

File: usaspending_api/references/v2/views/cfda.py
Comment (tony-sappe) [inline usaspending_api/references/v2/views/cfda.py]: endpoint doc string is wrong

File: usaspending_api/references/v2/views/cfda.py
Comment (tony-sappe) [inline usaspending_api/references/v2/views/cfda.py:42]: Wouldn't it be helpful to store the "cfdas" as the dictionary for looking up a specific CFDA? I would still be easy to  returning all CFDA records using `.values()`

File: usaspending_api/references/v2/views/cfda.py
Comment (tony-sappe) [inline usaspending_api/references/v2/views/cfda.py:24]: Since this is a GET request, shouldn't requesting a specific cfda which doesn't exist result in a 404 instead of an empty results response?

File: usaspending_api/settings.py
Comment (tony-sappe) [inline usaspending_api/settings.py:126]: Make sure you work with OPS on this, the older pattern was to not use env vars but inject the config into the settings,py file using regex. I never learned why....

File: usaspending_api/references/v2/views/cfda.py
Comment (BrianZito) [reply usaspending_api/references/v2/views/cfda.py]: This was probably updated while you were reviewing.

File: usaspending_api/references/v2/views/cfda.py
Comment (BrianZito) [reply usaspending_api/references/v2/views/cfda.py:24]: If the endpoint was always returning a single result instead of a list, I would absolutely agree with you. With there being a list (that could have multiple values), I think there's a grey area. An alternative would be to have the endpoint not return a list when filtered down to one CFDA code (since this should only be getting one result anyway), then returning 404 if the CFDA isn't found, but that would be changing the contract underneath frontend mid-sprint, so I'd  want to talk to John first.

File: usaspending_api/references/v2/views/cfda.py
Comment (BrianZito) [reply usaspending_api/references/v2/views/cfda.py:38]: As for the infinite loop, yes, it would be possible if the grants API went down indefinitely, although our daily redeploy means that no request would hog resources permanently. Either way, it would be easy enough to add a counter to just "give up" after a minute since we wouldn't be returning anything to the user after that point anyway.

It is true that if Grants API gives a different response structure than what was expected, this code would throw an exception. In that case, however, I would argue that the correct response would be to fail, and 500 would be the most appropriate response code, since a service used on the server side of the architecture failed.

File: usaspending_api/references/v2/views/cfda.py
Comment (tony-sappe) [reply usaspending_api/references/v2/views/cfda.py:24]: There are two endpoints in this view, the responses don't need to be identical between them.

Can multiple CFDAs be returned when the client provides a specific CFDA `/api/v2/references/cfda/totals/{?cfda_code}/` ?

Comment (BrianZito) [reply usaspending_api/references/v2/views/cfda.py:24]: It never should, since Grants API returns a JSON object instead of a JSON array, which guarantees unique keys. The only reason to keep this as a list is to avoid adding more work to @jonhill13 late in the sprint. John, do you forsee a problem if the api only returned results for a single CFDA instead of a results array when filtering on a single CFDA? I can get you a sample of what that would look like for you to develop against.

File: usaspending_api/references/v2/views/cfda.py
Comment (tony-sappe) [reply usaspending_api/references/v2/views/cfda.py:38]: > Additionally, it seems there is an expectation an exception will be raised in _request_from_grants_api() from .json()["cfdas"] since that is making hard assumptions about the response.

The context of that comment was for breaking out of the infinite loop. There was no explicit checking or error handling to help end the infinite loop.

In regards to your comment, it would be a good idea to add helpful logging to know if the issue was unreachable server, unexpected API response, authentication/missing api, or whatever likely situation you expect.

File: usaspending_api/settings.py
Comment (bminahan-kc) [reply usaspending_api/settings.py:126]: we can make this work. We will handle the same way as the database secrets. We should probalby rename the uwsgi file containing secrets from `uwsgi_db.ini` to something else now that there is a secret that is _not_ related to the databsae.

File: usaspending_api/references/v2/views/cfda.py
Comment (jonhill13) [reply usaspending_api/references/v2/views/cfda.py:26]: Yes, i would recommend returning just an object when returning a specific result and if an error or any other message is returned from the grants api return that.

The only reason I wrote the contract like that was to make it easier on the backend, but in hindsight, I do not know why I did that. I guess trying to be nice to the backend but we should be following best practice @BrianZito @tony-sappe

File: usaspending_api/settings.py
Comment (bminahan-kc) [reply usaspending_api/settings.py:126]: [this PR](https://github.com/fedspendingtransparency/data-act-build-tools/pull/269) will make sure the API key is added. Once that is merged in we should be good to go there.

**Review by jonhill13 — APPROVED** [review]

**Review by dpb-bah — APPROVED** [review]

**Comment by dpb-bah** [comment]
Approving but with a couple caveats:
* https://github.com/fedspendingtransparency/data-act-build-tools/pull/269 gets merged
* There's an underlying flaw with this approach in that it makes the endpoint dependent on an external API which could go down for any reason for any length of time out of our control. Suggestion would be to make a couple tickets to rework this on the backend such that 1) a nightly loader gathers the data needed from the Grants API and stores it in a table/view and 2) the endpoint the FE calls simply refers to that table/view.

**Comment by tony-sappe** [comment]
> Approving but with a couple caveats:
> 
> * [fedspendingtransparency/data-act-build-tools#269](https://github.com/fedspendingtransparency/data-act-build-tools/pull/269) gets merged
> * There's an underlying flaw with this approach in that it makes the endpoint dependent on an external API which could go down for any reason for any length of time out of our control. Suggestion would be to make a couple tickets to rework this on the backend such that 1) a nightly loader gathers the data needed from the Grants API and stores it in a table/view and 2) the endpoint the FE calls simply refers to that table/view.
>   @tony-sappe @jonhill13

@dpb-bah Thanks for reviewing the PR. There were a few discussions on the approach and when the sprint started it was expected the website react app would call the grants API directly. There were last-minute changes so this was settled on to complete the ticket. It's definitely a weakness, not sure I would call it a design flaw.

**Pull request merged**
